sys/riscv/sbi.c

```diff
@@ -0,0 +1,346 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/* MODIFIED */
+
+#include <sys/klog.h>
+#include <sys/mimiker.h>
+#include <riscv/sbi.h>
+
+/* SBI Implementation-Specific Definitions */
+#define OPENSBI_VERSION_MAJOR_OFFSET 16
+#define OPENSBI_VERSION_MINOR_MASK 0xFFFF
+
+typedef struct sbi_ret {
+  long error;
+  long value;
+} sbi_ret_t;
+
+/* SBI implementation info. */
+static u_long sbi_spec_version;
+static u_long sbi_impl_id;
+static u_long sbi_impl_version;
+
+static bool has_time_extension = false;
+static bool has_ipi_extension = false;
+static bool has_rfnc_extension = false;
+static bool has_srst_extension = false;
+static bool has_console_putchar_extension = false;
+static bool has_console_getchar_extension = false;
+
+/* Hardware implementation info. */
+static u_long mvendorid; /* CPU's JEDEC vendor ID */
+static u_long marchid;
+static u_long mimpid;
+
+static sbi_ret_t sbi_call(u_long ext, u_long func, u_long arg0, u_long arg1,
+                          u_long arg2, u_long arg3, u_long arg4) {
+  register register_t a0 __asm("a0") = (register_t)(arg0);
+  register register_t a1 __asm("a1") = (register_t)(arg1);
+  register register_t a2 __asm("a2") = (register_t)(arg2);
+  register register_t a3 __asm("a3") = (register_t)(arg3);
+  register register_t a4 __asm("a4") = (register_t)(arg4);
+  register register_t a6 __asm("a6") = (register_t)(func);
+  register register_t a7 __asm("a7") = (register_t)(ext);
+
+  __asm __volatile("ecall"
+                   : "+r"(a0), "+r"(a1)
+                   : "r"(a2), "r"(a3), "r"(a4), "r"(a6), "r"(a7)
+                   : "memory");
+
+  return (sbi_ret_t){
+    .error = a0,
+    .value = a1,
+  };
+}
```

Comment:
I'm not sure if we can trust compiler enough to use that tricky syntax - from my point of view it's better to write this function in pure assembler.

Reply:
FreeBSD and Linux do it this way, however, I'm not sure whether it requires some additional compiler options to handle it properly in all scenarios (e.g. when the -O3 is used). I have changed it to relay on bare assembly.